docs/adr/0005-runtime-compute-slashing.md

```diff
@@ -0,0 +1,284 @@
+
+    IORoot       *hash.Hash `json:"io_root,omitempty"`
+    StateRoot    *hash.Hash `json:"state_root,omitempty"`
+    MessageHash  *hash.Hash `json:"messages_hash,omitempty"`
+}
+
+type ExecutorCommitmentBody struct {
+  Header  ExecutorCommitmentHeader  `json:"header"`
+  Failure ExecutorCommitmentFailure `json:"failure,omitempty"`
+
+  TxnSchedSig      signature.Signature   `json:"txn_sched_sig"`
+  InputRoot        hash.Hash             `json:"input_root"`
+  InputStorageSigs []signature.Signature `json:"input_storage_sigs"`
+
+  // Optional fields (may be absent for failure indication).
+
+  StorageSignatures []signature.Signature   `json:"storage_signatures,omitempty"`
+  RakSig            *signature.RawSignature `json:"rak_sig,omitempty"`
+}
+```
+
+The notion of an _failure-indicating_ executor commitment is introduced as being
+an executor commitment with the following field values:
+
+- The `.failure` field must be present and non-zero. The code can indicate a
+  reason for the failure but currently the reason is ignored during processing.
+
+- `.header.round`, `.header.previous_hash`, `.txn_sched_sig`, `.input_root` and
+  `.input_storage_sigs` are set as for usual commitments (e.g., they must be
+  valid).
+
+- All other fields must be omitted or set to nil.
+
+### Root Hash Commitment Processing
+
+The processing of executor commitments by the commitment pool is modified as
+follows:
+
+- **Adding new commitments (`AddExecutorCommitment`)**
+
+  - If a commitment for a node already exists the existing commitment is
+    checked for evidence of double signing. Any evidence of misbehavior is
+    processed as described in the _Evidence_ subsection below.
+
+- **Discrepancy detection (`DetectDiscrepancy`)**
+
+  - If any executor commitment indicates failure, the discrepancy detection
+    process signals a discrepancy (which implies that discrepancy resolution is
+    triggered).
+
+- **Discrepancy resolution (`ResolveDiscrepancy`)**
+
+  - When tallying votes, any executor commitments indicating failure are tallied
+    into its own bucket. If the failure bucket receives 1/2+ votes, the round
+    fails.
+
+  - If after discrepancy resolution a non-failure option receives 1/2+ votes,
+    this is considered the correct result. Executor commitments for any other
+    result (excluding failure indication) are considered incorrect and are
+    subject to slashing (based on the configured slashing instructions for the
+    `SlashRuntimeIncorrectResults` reason).
+
+A portion of slashed funds is disbursed equally to the compute nodes which
+participated in discrepancy resolution for the round. The remainder of slashed
+funds is transferred to the runtime account.
+
+Any slashing instructions related to freezing nodes are currently ignored.
+
+### Transaction Methods
+
+This proposal updates the following transaction methods in the roothash module:
+
+#### Evidence
+
+The evidence method allows anyone to submit evidence of runtime node
```

Comment:
Anyone can submit evidence of double signing, but there's no evidence format to be committed to support discrepancy resolution based slashing.  Do we want/need this?